hal_aarch64/src/irq.rs

```diff
@@ -6,185 +6,210 @@ use hal_core::{Error, TimerCallbackFn};
-    super::cpu::unmask_interrupts();
-
-    Ok(())
-}
-
-enum IrqChip {
-    NoChip,
-    GicV2(GicV2),
-}
-
-impl IrqChip {
-    fn get_int(&mut self) -> Result<u32, Error> {
-        match self {
-            Self::NoChip => unreachable!("does not support this"),
-            Self::GicV2(gic) => gic.get_int(),
-        }
-    }
-
-    fn clear_int(&mut self, int: u32) {
-        match self {
-            Self::NoChip => unreachable!("does not support this"),
-            Self::GicV2(gic) => gic.clear_int(int),
+macro_rules! gen_isr_stub {
+    () => {
+        concat!(
+            r#"
+            .balign 0x80
+            msr spsel, xzr
+            stp x0, x1, [sp, #-16]!
+            stp x2, x3, [sp, #-16]!
+            stp x4, x5, [sp, #-16]!
+            stp x6, x7, [sp, #-16]!
+            stp x8, x9, [sp, #-16]!
+            stp x10, x11, [sp, #-16]!
+            stp x12, x13, [sp, #-16]!
+            stp x14, x15, [sp, #-16]!
+            stp x16, x17, [sp, #-16]!
+            stp x18, x29, [sp, #-16]!
+            stp x30, xzr, [sp, #-16]!
+
+            mov x0, . - el1_vector_table
+            bl aarch64_common_trap
+
+            ldp x30, xzr, [sp], #16
+            ldp x18, x29, [sp], #16
+            ldp x16, x17, [sp], #16
+            ldp x14, x15, [sp], #16
+            ldp x12, x13, [sp], #16
+            ldp x10, x11, [sp], #16
+            ldp x8, x9, [sp], #16
+            ldp x6, x7, [sp], #16
+            ldp x4, x5, [sp], #16
+            ldp x2, x3, [sp], #16
+            ldp x0, x1, [sp], #16
+            eret
+            "#
+        )
+    };
+}
+
+#[naked]
+#[no_mangle]
+#[repr(align(0x800))]
+unsafe extern "C" fn el1_vector_table() {
+    core::arch::asm!(
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        gen_isr_stub!(),
+        options(noreturn)
+    );
+}
+
+#[repr(u64)]
+#[derive(Debug)]
+enum InterruptType {
+    // Current EL with SP0
+    SyncCurrentElSp0,
+    IrqCurrentElSp0,
+    FiqCurrentElSp0,
+    SerrorCurrentElSp0,
+    // Current EL with SPx
+    SyncCurrentElSpx,
+    IrqCurrentElSpx,
+    FiqCurrentElSpx,
+    SerrorCurrentElSpx,
+    // Lower EL
+    SyncLowerEl,
+    IrqLowerEl,
+    FiqLowerEl,
+    SerrorLowerEl,
+    // Lower EL with aarch32
+    SyncLowerElAarch32,
+    IrqLowerElAarch32,
+    FiqLowerElAarch32,
+    SerrorLowerElAarch32,
+}
+
+/// Need a once lock because this is unset until Aarch64Irqs::init has been called by the hal.
+static IRQS: OnceLock<&Aarch64Irqs> = OnceLock::new();
+
+#[no_mangle]
+unsafe extern "C" fn aarch64_common_trap(offset: u64) {
+    log::debug!("aarch64_common_trap(0x{:x})", offset);
+
+    let int_type = match offset {
+        0x000..=0x07f => InterruptType::SyncCurrentElSp0,
```

Comment:
note that since Rustc 1.80 we can use exclusive range in patterns, so if it is easier this can be written as 

```suggestion
        0x000..0x080 => InterruptType::SyncCurrentElSp0,
```

don't know if the spec uses exclusive limits or not so feel free to ignore

Reply:
Iirc I was following the spec